Sources/MapboxMobileEvents/MMELocationManager.m

```diff
@@ -37,32 +37,46 @@ - (void)dealloc {
     _locationManager.delegate = nil;
 }
 
-- (CLLocationManager *)locationManager  {
-    @synchronized (self) {
-        if (!_locationManager) {
-            dispatch_block_t instantiateLocationManager = ^{
-                _locationManager = [[MMEDependencyManager sharedManager] locationManagerInstance];
-            };
-            if ([NSThread isMainThread]) {
-                instantiateLocationManager();
-            } else {
-                dispatch_sync(dispatch_get_main_queue(), instantiateLocationManager);
-            }
-        }
-        return _locationManager;
+- (CLLocationManager *)locationManager {
+    // Always reschedule to main queue to avoid race condition
+    if (![NSThread isMainThread]) {
+        __block CLLocationManager* manager = nil;
+        // Avoid other synhronization methods like @synchronized
+        // since the access to locationManager might happen in
+        // main thread after this function call but before
+        // mainQueue async.
+        dispatch_sync(dispatch_get_main_queue(), ^{
```

Comment:
hm, you are right. Any ideas on how to fix that? 🙂

Reply:
I think it is acceptable risk. My point is that `sync` call from main thread to the background queue in not something common and should be generally avoided as a bad practice leading to the UI lags.

Reply:
Thanks for providing a workaround. It seems to have some overhead so I will avoid it at the moment. In case of any issues reports I will reconsider my position.